### PR TITLE
add cgo support

### PR DESCRIPTION
**Pull request by mikedanese**
last two commits.

### CODE REVIEW

**Comment by rmmh** [comment]
t

File: gazel.go [hunk 1]
```diff
@@ -217,58 +214,119 @@ func (v *Vendorer) updateSinglePkg(path string) error {
 	return v.updatePkg(path, "", pkg)
 }
 
+type RuleType int
+
+const (
+	RuleTypeGoBinary RuleType = iota
+	RuleTypeGoLibrary
+	RuleTypeGoTest
+	RuleTypeGoXTest
```
Comment (ixdy) [inline gazel.go:223]: what is an XTest?

**Comment by spxtr** [comment]
I don't see anything offensive here, but it's pretty hard to follow this code.

**Comment by mikedanese** [comment]
This was merged in the gazel subdir

File: gazel.go [hunk 2]
```diff
@@ -217,58 +214,119 @@ func (v *Vendorer) updateSinglePkg(path string) error {
...
 	if pkg.IsCommand() {
-		rules = append(rules, newRule("go_binary", filepath.Base(pkg.Dir), attrs))
+		rules = append(rules, newRule(RuleTypeGoBinary, namer, attrs))
 	} else {
-		rules = append(rules, newRule("go_library", "go_default_library", attrs))
+		addGoDefaultLibrary := len(cgoSrcs.List) > 0 || len(srcs.List) > 0
+		if len(cgoSrcs.List) != 0 {
+			cgoRule := newRule(RuleTypeCGoGenrule, namer, map[string]bzl.Expr{
+				"srcs":      cgoSrcs,
+				"clinkopts": asExpr([]string{"-lz", "-lm", "-lpthread", "-ldl"}),
```
Comment (ixdy) [inline gazel.go:298]: I think I mentioned this before, but this list seems somewhat arbitrary, so documenting where it came from is probably a good idea.

File: gazel.go [hunk 3]
```diff
@@ -217,58 +214,119 @@ func (v *Vendorer) updateSinglePkg(path string) error {
...
+			rules = append(rules, cgoRule)
+			attrs["library"] = asExpr(namer(RuleTypeCGoGenrule))
+		}
 		if len(pkg.TestGoFiles) != 0 {
-			rules = append(rules, newRule("go_test", "go_default_test", map[string]bzl.Expr{
-				"srcs":    asExpr(pkg.TestGoFiles),
-				"deps":    v.extractDeps(pkg.TestImports),
-				"library": asExpr("go_default_library"),
-			}))
+			xtestRule := newRule(RuleTypeGoTest, namer, map[string]bzl.Expr{
```
Comment (ixdy) [inline gazel.go:304]: now I'm even more confused. why is this named `xtestRule` if it's a GoTest, not GoXTest (whatever that is)?

File: gazel.go [hunk 4]
```diff
@@ -506,6 +563,34 @@ func ReconcileRules(path string, rules []*bzl.Rule, dryRun bool) (bool, error) {
...
+	for _, r := range f.Rules("load") {
+		const goRulesLabel = "@io_bazel_rules_go//go:def.bzl"
+		args := bzl.Strings(&bzl.ListExpr{List: r.Call.List})
+		if len(args) == 0 {
+			continue
+		}
+		if args[0] != goRulesLabel {
+			continue
+		}
+		r.Call.List = asExpr(append(
```
Comment (ixdy) [inline gazel.go:587]: how does this work for BUILD files which manually extend the `load()` rule? for example, the `BUILD` file in the kubernetes root loads `go_prefix`.

**Review by ixdy — COMMENTED** [review]
I know you already merged this, but I have some thoughts. :)

I agree with @spxtr that this is a bit hard to read.

tests might be nice, too. :)

File: gazel.go (hunk 1 above)
Comment (mikedanese) [reply gazel.go:223]: it stands for external test. You can have a test in package foo and have your tests be in packge foo_test. We do that in various places like https://github.com/kubernetes/kubernetes/blob/master/pkg/api/meta_test.go#L17

File: gazel.go (hunk 4 above)
Comment (mikedanese) [reply gazel.go:587]: You can load("@io_bazel_rules_go//go:def.bzl") later on in the file and only the first load() is managed. a bit hacky i know, but open to alternatives.

File: gazel.go (hunk 2 above)
Comment (mikedanese) [reply gazel.go:298]: I don't remember but am looking...

File: gazel.go (hunk 3 above)
Comment (mikedanese) [reply gazel.go:304]: xtests and regular tests both use RuleTypeGoTest (they are both just go_test targets).

File: gazel.go (hunk 4 above)
Comment (ixdy) [reply gazel.go:587]: is there some way we can inject a comment like "# Automanaged" above the automanaged `load` rule?

I see https://github.com/bazelbuild/buildifier/blob/251fa7607cb9da4c9b3505af634ae1e11517d987/core/syntax.go#L55-L81 but I'm not yet sure how to use it.

File: gazel.go (hunk 3 above)
Comment (ixdy) [reply gazel.go:304]: yes, I'm just confused as to why this is a `RuleTypeGoTest` but the variable is named `xtestRule` rather than `testRule`.
